common/namespace/registry.go

```diff
@@ -483,6 +489,149 @@ func (r *registry) getNamespaceByIDLocked(id ID) (*Namespace, error) {
 	return nil, serviceerror.NewNamespaceNotFound(id.String())
 }
 
+// getOrPutNamespace retrieves the information from the cache if it exists or reads through
+// to the persistence layer and updates caches if it doesn't
+func (r *registry) getOrPutNamespace(name Name) (*Namespace, error) {
+	// check main caches
+	cacheHit, _ := r.getNamespace(name)
+	if cacheHit != nil {
+		return cacheHit, nil
+	}
+
+	// check readthrough cache to prevent excessive requests to persistence layer
+	prevErr := r.getPreviousReadthroughError(name.String())
+	if prevErr != nil {
+		return nil, prevErr
+	}
+
+	// readthrough to persistence layer and update readthrough cache if not found
+	ns, err := r.getNamespaceByNamePersistence(name)
+	if err != nil {
+		return nil, err
+	}
+
+	// update main caches if found
+	r.updateCachesSingleNamespace(ns)
+
+	return ns, nil
+}
+
+// getOrPutNamespaceByID retrieves the information from the cache if it exists or reads through
+// to the persistence layer and updates caches if it doesn't
+func (r *registry) getOrPutNamespaceByID(id ID) (*Namespace, error) {
+	// check main caches
+	cacheHit, _ := r.getNamespaceByID(id)
+	if cacheHit != nil {
+		return cacheHit, nil
+	}
+
+	// check readthrough cache to prevent excessive requests to persistence layer
+	prevErr := r.getPreviousReadthroughError(id.String())
+	if prevErr != nil {
+		return nil, prevErr
+	}
+
+	// readthrough to persistence layer and update readthrough cache if not found
+	ns, err := r.getNamespaceByIDPersistence(id)
```

Comment:
Concurrent get namespace request can issue multiple persistence requests?

I am thinking for one ns (for even only all ns) there should be only one pending persistence request. When that request is pending, other read through requests should be blocked.